icefoganalytics/travel-authorization
# Add Expense Read View for Manager Type With Custom Manager Actions

Fixes https://github.com/icefoganalytics/travel-authorization/issues/149

Relates to:
- [Analysis, Expense Pages, 2023-12-19](https://docs.google.com/document/d/1cWEA-0w9ro1TyajfnffkYaKL42yYnA89i9gdQysht_A/edit?pli=1#heading=h.zc2nwn7nmjlu)

# Context

**Is your feature request related to a problem? Please describe.**

Add read-only expense tab for the generic “manager” type.
Add special manager only actions for expenses.
- reassign
- review - this might be for expenses?
- verify (finanace) - needs more shaping
- approve - not sure if this was for submitted status, and isn't relevant here?
- pay - needs more shaping

**Describe the solution you'd like**

Needs to have the various abilities seen in this wireframe, though not likely in this form.
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/88ddd432-ca6c-44ce-afaf-739621d67632)

# Implementation

Genericize a ReadMyTravelAuthorization and ManageTravelAuthorization components so they can share read-only components.
Add Read (my) and Manage (any) views for travel authorizations.
Add useCurrentUser global store and reset it on logout/signin.
Rename `expense_claim` status to `expense_claim_submitted`
Add `expense_claim_approved` and `expense_claim_denied` states.

## Suggestions

Tweak the readonly expense view and merge the expense and meals and incidentals tables.

# Screenshots

Manage Expense tab
![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/e96bd806-07ff-4613-ab4d-832045be8c77)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the "My Travel Requests" page via the top drop down nav.
5. Create a new travel authorization via the "+ Travel Authorization" page.
6. Fill in the Round Trip section making sure that you pick dates that are in the past.
7. Scroll down and generate and estimate via the call-to-action button.
8. Submit the travel authorization to yourself.
9. Go to the "Manager View" via the top drop down nav.
10. In the "Pending Approvals" section click on the latest travel authorization.
11. Scroll to the bottom and click the "Approve" button.
12. Go back to the "My Travel Requests" page via the top drop down nav.
13. Find and click on your latest travel request, now approved. It might be on the second page.
14. Go to the Expense tab.
15. Click on the Prefill Expenses call-to-action.
16. Add receipts to all of your expenses.
17. Scroll down and add some coding rows.
18. Submit the expense page to your supervisor.
19. Note that you are redirected to the read only expense tab view, and its now fully built out.
20. Go to the "Manager View" via the top drop down nav, again.
21. Find your travel authorization request in the "Awaiting Expense Approval" table and click on it.
22. Go to the "Expense" tab and scroll down to the re-assignment action.
23. Reassign the request to another user.
24. Note that this bounces you out of the page, as reassignment might have lost you access to the travel request.
25. Do some database mudging to regain access to the expense manage page again, or repeat the previous steps with a new travel authorization.
26. Scroll down to the approve/deny buttons.
27. Approve" the travel request expenses. Sets state to expense_claim_approved.
28. Repeat everything, but this time deny the request.
